CIPs/0005.md

```diff
@@ -0,0 +1,74 @@
+2) Maintain a healthy overlay network for IBFT.   This is to maximize consistent block times, which obviously has direct impact to the consumer end users.
+
+This CIP is an iteration of the Announce protocol that is implemented in https://github.com/celo-org/celo-blockchain/pull/816.  The original protocol does optimize for 2), but not for 1).
+
+## Goals
+
+- Reduce bandwidth requirements of the full nodes.
+- Maximize availability of the IBFT overlay network.
+
+## Proposed Solution
+
+The announce protocol has to be able to deal with two fundamental faulty scenarios.  For a given validator pair, when only one of the validators has the correct enodeURL of the other (the asymmetric ignorance case) and when both of the validators don't have the correct enodeURL of the other (the mutual ignorance case).
+
+To deal with these cases, this proposal proposes adding the following high level components to the existing Announce protocol.  1) Add the capability for a validator to directly communicate to another remote validator it's assigned enodeURL, 2) add fast discovery of validator enode mapping changes, and 3) add capability for a validator to query another validator for it's assigned enodeURL via a gossiped message.  Details of these components are in the subsections below.
+
+The asymmetric ignorance case can then be handled by having the validators directly communicate to each remote validator it's assigned enodeURL whenever the enodeURL mapping is updated.  Note that this will need to happen for all remote validators, even if the assigned URL is not updated for a subset of the validators, since all validators will need to know what is the updated version number.  Also, this direct communication will need to be handled when there is and is not an existing connection between validators.
+
+The mutual ignorance case can be handled by having each validator monitor when a validator enode mapping changes, and when it realizes that it's changed, it will gossip a query message to the validator for the updated assigned enodeURL.  That validator will then directly communicate it to the querying validator.
+
+### Direct communication of assigned enodeURL
+
+This component will include the following changes:
+
+1)  Announce message format will need to be changed so that only one authenticated and encrypted enodeURL (along with the version number) can be put in the message payload
+2)  On enodeURL mapping update, that validator will need to directly send an announce message to all validators. In the case of when the validator already has a connection with the remote validator, the originating validator should simply send an announce message to it.  In the case when that connection doesn't exist, then the announce message will need to be sent as part of the connection establishing handshake.
```

Comment:
What does enodeURL mapping update mean?

Reply:
What is the benefit of directly sending the announce message? Purely speed? There are no actual bandwidth savings, right?

Reply:
Will a direct announce receiver gossip the announce further?

Reply:
> What does enodeURL mapping update mean?
Each validator will maintain a mapping of remote validator -> enodeURL that they should use to connect to the originating validator.  Whenever that mapping is updated, then the originating validator will gossip out an updated version number.






> What is the benefit of directly sending the announce message? Purely speed? There are no actual bandwidth savings, right?

There actually will be a ton of bandwidth savings for full nodes, since the announce message will not be gossiped throughout the network.



> Will a direct announce receiver gossip the announce further?

No, the receiver will not.